### PR TITLE
Update Array to respect FSStore's key_separator 

### LINKED ISSUE #717: Array._chunk_key does not respect FSStore's key_separator attribute
#### Minimal, reproducible code sample, a copy-pastable example if possible

```Python
data = np.arange(1000).reshape(10, 10, 10)
name = 'raw'

store = FSStore('/tmp/test.zr', key_separator='/', auto_mkdir=True)
f = open_group(store, mode='w')
f.create_dataset(name, data=data, chunks=(5, 5, 5),
                 compressor=None)
h = open_group(store, mode='r')
# check below fails because h[name][:] will be all zeros
np.testing.assert_array_equal(h[name][:], data)  
```

#### Problem description

If an nested Zarr array is read in using an FSStore object initialized with `key_separator='/'`, all array entries are reported as zeros.

This seems to be because the `_chunk_key` attribute of `zarr.core.Array` assumes that '.' is the separator

#### Version and installation information

'2.7.1.dev2' (current master branch)

------

I will make a PR addressing this now

### PR DESCRIPTION
**Pull request by grlee77**
[Description of PR]

closes #717

I added a couple of FSStore test classes to test_hierarchy.py. The "NestedFSStore" class tests the fix on the specific case raised in #717.

TODO:
* [x] Add unit tests and/or doctests in docstrings
~* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions~
~* [ ] New/modified features documented in docs/tutorial.rst~
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: zarr/core.py [hunk 1]
```diff
@@ -1950,7 +1950,11 @@ def _process_for_setitem(self, ckey, chunk_selection, value, fields=None):
         return self._encode_chunk(chunk)
 
     def _chunk_key(self, chunk_coords):
-        return self._key_prefix + '.'.join(map(str, chunk_coords))
+        if hasattr(self._store, 'key_separator'):
```
Comment (joshmoore) [inline zarr/core.py:1953]: A heads up that #716 deprecates `key_separator` in favor `store._dimension_separator`. Looking at this usage, I imagine wrapping stores like `LRUCacheStore` and `ConsolidatedMetadataStore` are going to need to set `_dimension_separator as well.

Comment (grlee77) [reply zarr/core.py:1953]: Okay, thanks for letting me know. Should we hold off on this PR until #716 is in and then revisit?

Comment (joshmoore) [reply zarr/core.py:1953]: Nah, let's get this out as a point release. Just a warning that this "API" will change. It's all kismet.

**Review by joshmoore — APPROVED** [review]

File: zarr/core.py (hunk 1 above)
Comment (d-v-b) [reply zarr/core.py:1953]: The behavior of `_chunk_key` should not be changed here. Instead, the behavior of `FSStore._getitems_` should be changed to properly remap the keys. See https://github.com/zarr-developers/zarr-python/pull/709/files#diff-565e487a2f60258b6baa2e4db8ef175cc16b8a949651834bd43d0a9f21e07358R1054

**Comment by d-v-b** [comment]
This PR needs to be reverted, since the solution in this PR breaks the abstraction boundary between array and chunk store, and the problem this PR solves is addressed in https://github.com/zarr-developers/zarr-python/pull/709 by changes to `FSStore`, which is the correct level of abstraction for this fix.

**Comment by joshmoore** [comment]
Thanks, @d-v-b. And sorry for the troubles. If you haven't already, can you capture the test from here in #709?

**Comment by joshmoore** [comment]
see https://github.com/zarr-developers/zarr-python/pull/729

**Pull request merged**